docs/src/index.md

```diff
@@ -6,4 +6,369 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
 
 [ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl) provides a variety of common utilities that can be used by downstream automatic differentiation (AD) tools to define and execute forward-, reverse-, and mixed-mode primitives.
 
-This package is a work-in-progress, as is the documentation. Contributions welcome!
+### Introduction:
+ChainRules is all about providing a rich set of rules for doing differentiation.
+When a person does introductory calculus, they learn that the derivative (with respect to `x`)
+of `a*x` is `a`, and the derivative of `sin(x)` is `cos(x)`, etc.
+And they learn how to combine simple rules, via the chain rule, to differentiate complicated functions.
+ChainRules.jl basically a progamatic repository of that knowledge, with the generalizations to higher dimensions.
+
+Autodiff (AD) tools roughly work by reducing a problem down to simple parts that they know the rules for,
+and then combining those rules.
+Knowing rules for more complicated functions speeds up the autodiff process as it doesn't have to break things down as much.
+
+** ChainRules is an AD independent collection of rules to use in an differentiation system **
+
+### `rrule` and `frule`
+
+!!! Terminology "`rrule` and `frule`"
+
+    `rrule` and `frule` are ChainRules.jl specific terms.
+    And there exact functioning is kind of ChainRule specific,
+    though other tools may do similar.
+    The core notion is sometimes called
+    _Custom AD primitives_, _custom adjoints_, _custom sensitivities_.
+
+The rules are encoded as `rrules` and `frules`,
+for use in reverse-mode and forward-mode differentiation respectively.
+
+The `rrule` for some function `foo`, takes the positional argument `args` and keyword argument `kwargs` is written:
+```julia
+function rrule(::typeof(foo), args; kwargs...)
+    ...
+    return y, pullback
+end
+```
+where `y` must be equal to `foo(args; kwargs...)`,
+and _pullback_ is a function to propagate the derivative information backwards at that point (more later).
+Often but not always it is calculated directly.
+The exception is that we can calculate it indirectly to make
```

Comment:
>The exception is that we can calculate it indirectly to make
the `pullback` faster. (more on _pullback_ later)

I don't know what this is trying to tell me.

Reply:
It needs an example 
Only example I have is

https://github.com/JuliaDiff/ChainRules.jl/pull/91/files#diff-e59f5e7b34490be11e2a82fdb8b831edR105-R123

Or the one in #104

Reply:
Seems a bit in the weeds here. Maybe bring this up in the `pullback` section?

Reply:
Might be better to put in the FAQ section,
like "Why does rrule/frule also calculate the main result" ?

I think that could be clearer, and there are 2 reasons.
1 is doing it differently to get intermidiary values to use in the propagator (which I mentioned),
the other is if the output is used in the backpropagator (like it is for deriviative of sigmoid).